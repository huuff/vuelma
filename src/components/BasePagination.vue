<template>
<nav class="pagination" role="navigation" aria-label="pagination">
  <a class="pagination-previous" aria-label="previous page">
    <font-awesome-icon icon="angle-left"/>
  </a>
  <a class="pagination-next" aria-label="next page">
    <font-awesome-icon icon="angle-right"/>
  </a>
  <template v-if="pageNumber > 0">
    <ul class="pagination-list">
      <base-pagination-link 
        v-if="!backwardReachesFirst"
        :pageNumber="1" 
      />

      <li v-if="!backwardReachesFirst">
        <span class="pagination-ellipsis">&hellip;</span>
      </li>
      
      <base-pagination-link v-for="page in showBackward"
        :pageNumber="page"
        :key="`page-${page}`"
      />

      <base-pagination-link
        :pageNumber="currentPage"
      />

      <base-pagination-link v-for="page in showForward"
        :pageNumber="page"
        :key="`page-${page}`"
      />

      <li v-if="!forwardReachesLast">
        <span class="pagination-ellipsis">&hellip;</span>
      </li>

      <base-pagination-link 
        v-if="!forwardReachesLast"
        :pageNumber="pageNumber"
      />
    </ul>
  </template>
</nav>
</template>

<script setup lang="ts">
// TODO: Provides and inject for emitting goto
// TODO: Validation that current page must be between 1 and last page?
// TODO: Obviously, goto page events!
// TODO: Test it!
import { computed, provide } from 'vue';
import BasePaginationLink from '@/components/BasePaginationLink.vue';
import { library } from '@fortawesome/fontawesome-svg-core';
import { faAngleRight, faAngleLeft } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome';

library.add(faAngleLeft);
library.add(faAngleRight);

const props = withDefaults(defineProps<{
  pageNumber: number;
  currentPage: number;
  showAround?: number;
}>(), {
  showAround: 2,
});

provide('currentPage', props.currentPage);

const showBackward = computed(() => {
  const first = Math.max(1, props.currentPage - props.showAround);
  const pagesToShow: number[] = [];

  for (let page = first; page < props.currentPage; page++)
    pagesToShow.push(page);
  
  return pagesToShow;
});

const showForward = computed(() => {
  const last = Math.min(props.pageNumber, props.currentPage + props.showAround + 1);
  const pagesToShow: number[] = [];

  for (let page = props.currentPage + 1; page <= last; page++)
    pagesToShow.push(page);

  return pagesToShow;
});

const backwardReachesFirst = computed(() => {
  return showBackward.value.includes(1) || showBackward.value.length === 0;
});

const forwardReachesLast = computed(() => {
  return showForward.value.includes(props.pageNumber) || showForward.value.length === 0;
});

</script>
