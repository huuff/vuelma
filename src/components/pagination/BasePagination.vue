<template>
<nav class="pagination" role="navigation" aria-label="pagination">
  <button
    type="button"
    :disabled="currentPage === 1"
    @click="emit('update:currentPage', currentPage-1)"
    class="pagination-previous button" 
    aria-label="previous page"
    ><font-awesome-icon icon="angle-left"/>
  </button>
  <button
    @click="emit('update:currentPage', currentPage+1)"
    :disabled="currentPage === pageNumber"
    type="button"
    class="pagination-next button "
    aria-label="next page"
  >
    <font-awesome-icon icon="angle-right"/>
  </button>
  <template v-if="pageNumber > 0">
    <ul class="pagination-list">
      <base-pagination-link 
        v-if="!backwardReachesFirst"
        :pageNumber="1" 
      />

      <li v-if="!backwardReachesFirst && !showBackward.includes(2)">
        <span class="pagination-ellipsis">&hellip;</span>
      </li>
      
      <base-pagination-link v-for="page in showBackward"
        :pageNumber="page"
        :key="`page-${page}`"
      />

      <base-pagination-link
        :pageNumber="currentPage"
      />

      <base-pagination-link v-for="page in showForward"
        :pageNumber="page"
        :key="`page-${page}`"
      />

      <li v-if="!forwardReachesLast && !showForward.includes(pageNumber - 1)">
        <span class="pagination-ellipsis">&hellip;</span>
      </li>

      <base-pagination-link 
        v-if="!forwardReachesLast"
        :pageNumber="pageNumber"
      />
    </ul>
  </template>
</nav>
</template>

<script setup lang="ts">
// XXX: Pretty shitty logic in here but I'm no mathematician
import { computed, } from 'vue';
import { provideAccessors } from '@/composables/injected-accessors'
import BasePaginationLink from '@/components/pagination/BasePaginationLink.vue';
import { library } from '@fortawesome/fontawesome-svg-core';
import { faAngleRight, faAngleLeft } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome';
import partial from 'lodash/partial';

library.add(faAngleLeft);
library.add(faAngleRight);

const props = withDefaults(defineProps<{
  pageNumber: number;
  currentPage: number;
  showAround?: number;
}>(), {
  showAround: 2,
});

const emit = defineEmits<{
  (event: 'update:currentPage', page: number): void
}>();

provideAccessors("CurrentPage", () => props.currentPage, partial(emit, "update:currentPage"));

const showBackward = computed(() => {
  const first = Math.max(1, props.currentPage - props.showAround);
  const pagesToShow: number[] = [];

  for (let page = first; page < props.currentPage; page++)
    pagesToShow.push(page);
  
  return pagesToShow;
});

const showForward = computed(() => {
  const last = Math.min(props.pageNumber, props.currentPage + props.showAround);
  const pagesToShow: number[] = [];

  for (let page = props.currentPage + 1; page <= last; page++)
    pagesToShow.push(page);

  return pagesToShow;
});

const backwardReachesFirst = computed(() => {
  return showBackward.value.includes(1) || showBackward.value.length === 0;
});

const forwardReachesLast = computed(() => {
  return showForward.value.includes(props.pageNumber) || showForward.value.length === 0;
});

if (props.currentPage < 1 || props.currentPage > props.pageNumber) {
  throw new Error(`Current page can't be over the pageNumber or under 1, current value: ${props.currentPage}`)
}

</script>
