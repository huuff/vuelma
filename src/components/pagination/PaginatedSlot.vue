<template>
<base-pagination v-model:currentPage="actualCurrentPage" :pageNumber="pageNumber" :showAround="showAround">
</base-pagination>
<div>
  <slot :items="itemsToShow"></slot> 
</div>
</template>

<script setup lang="ts">
// TODO: For some reason, it starts out with a different number of pages
import { toRef, computed } from "vue";
import BasePagination from "./BasePagination.vue";
import { useOptionalTwoWayBinding } from "@/composables/optional-two-way-binding";
import partial from "lodash/partial";

const props = withDefaults(defineProps<{
  items: unknown[];
  itemsPerPage?: number;
  currentPage?: number;
  showAround?: number;
}>(), {
  showAround: 2,
  currentPage: undefined,
  itemsPerPage: 10,
});

const emit = defineEmits<{
  (event: "update:currentPage", newCurrentPage: number): void;
}>();

const actualCurrentPage = useOptionalTwoWayBinding(1, toRef(props, "currentPage"), partial(emit, "update:currentPage"));

const pageNumber = computed(() => Math.floor(props.items.length / props.itemsPerPage));

const itemsToShow = computed(() => {
  const start = pageNumber.value * props.itemsPerPage;
  return props.items.slice(start, start + props.itemsPerPage);
});
</script>
