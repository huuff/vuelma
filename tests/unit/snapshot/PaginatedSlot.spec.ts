import { mount } from "@vue/test-utils";
import BasePagination from "@/components/pagination/BasePagination.vue";
import PaginatedSlot from "@/components/pagination/PaginatedSlot.vue";

test("PaginatedSlot.vue snapshot", () => {
  const wrapper = mount(PaginatedSlot, {
    props: {
      currentPage: 5,
      itemsPerPage: 10,
      showAround: 2,
      items: Array.from(Array(100).keys()),
    },
    global: {
      components: { BasePagination }
    },
    slots: {
      default: `
        <template #default="{items}">
          <ul>
            <li v-for="item in items" :key="item.toString()">Item {{item}}</li>
          </ul>
        </template>
      `,
    },
  });

  expect(wrapper.element).toMatchSnapshot();
});
