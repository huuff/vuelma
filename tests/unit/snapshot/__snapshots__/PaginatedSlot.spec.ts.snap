// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginatedSlot.vue snapshot 1`] = `
<div
  data-v-app=""
>
  
  <nav
    aria-label="pagination"
    class="pagination"
    role="pagination"
  >
    <button
      aria-label="previous"
      class="pagination-previous"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-angle-left fa-w-8"
        data-icon="angle-left"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 256 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          class=""
          d="M31.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L127.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L201.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34z"
          fill="currentColor"
        />
      </svg>
    </button>
    <button
      aria-label="next"
      class="pagination-next"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-angle-right fa-w-8"
        data-icon="angle-right"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 256 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          class=""
          d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
          fill="currentColor"
        />
      </svg>
    </button>
    <ul
      class="pagination-list"
    >
      <li>
        <a
          aria-label="Goto page 1"
          class="pagination-link"
        >
          1
        </a>
      </li>
      <li>
        <a
          class="pagination-ellipsis"
        >
          …
        </a>
      </li>
      
      <li>
        <a
          aria-label="Goto page 3"
          class="pagination-link"
        >
          3
        </a>
      </li>
      <li>
        <a
          aria-label="Goto page 4"
          class="pagination-link"
        >
          4
        </a>
      </li>
      
      <li>
        <a
          aria-current="true"
          aria-label="Goto page 5"
          class="pagination-link is-current"
        >
          5
        </a>
      </li>
      
      <li>
        <a
          aria-label="Goto page 6"
          class="pagination-link"
        >
          6
        </a>
      </li>
      <li>
        <a
          aria-label="Goto page 7"
          class="pagination-link"
        >
          7
        </a>
      </li>
      
      <li>
        <a
          class="pagination-ellipsis"
        >
          …
        </a>
      </li>
      <li>
        <a
          aria-label="Goto page 10"
          class="pagination-link"
        >
          10
        </a>
      </li>
    </ul>
  </nav>
  <div>
    
    <ul>
      
      <li>
        Item 40
      </li>
      <li>
        Item 41
      </li>
      <li>
        Item 42
      </li>
      <li>
        Item 43
      </li>
      <li>
        Item 44
      </li>
      <li>
        Item 45
      </li>
      <li>
        Item 46
      </li>
      <li>
        Item 47
      </li>
      <li>
        Item 48
      </li>
      <li>
        Item 49
      </li>
      
    </ul>
    
  </div>
  
</div>
`;
